src/org/omegat/util/StringUtil.java

```diff
@@ -511,4 +512,235 @@ public static String format(String str, Object... arguments) {
         str = str.replaceAll("'", "''");
         return MessageFormat.format(str, arguments);
     }
+
+    /**
+     * create match expression for full/half width character
+     *
+     * Analyze text whether contains full/half width character
+     * that has a codepoint in both full-width and half-width expression in Unicode
+     *    ex. ASCII, full-width alphabet, full-width katakana, half-width katakana, etc.
+     *
+     * @param  String
+     * @return String expresssion such as '(A|\uff21)(B|\uff22)(C|\uff23)'
+     **/
+    public static String createFullHalfMatchExpression(String text) {
+        StringBuilder sb = new StringBuilder(text);
+        StringBuilder result = new StringBuilder();
+
+        for (int cp, i = 0; i < sb.length(); i += Character.charCount(cp) ) {
+            cp = sb.codePointAt(i);
+            // ASCII
+            if (( cp >= 0x0021 ) && ( cp <= 0x007E )) {
+                result.append(genFullHalfMatchRegex(quote(cp), (char)(cp+0xFEE0)));
+                continue;
+            }
+            // Half width Hangul
+            if (( cp > 0xFFA1 ) && ( cp <= 0xFFBE )) {
+                result.append(genFullHalfMatchRegex((char)cp, (char)(cp-0xCE70)));
+                continue;
+            }
+            // Full width alphabet
+            if (0xff01 <  cp &&  cp < 0xff5e) {
+                result.append(genFullHalfMatchRegex((char)cp, quote(cp-0xFEE0)));
+                continue;
+            }
+            // Full width Hangul
+            if (( cp > 0x3131 ) && ( cp <= 0x314E )) {
+                result.append(genFullHalfMatchRegex((char)cp, (char)(cp+0xCE70)));
+                continue;
+            }
+            switch ( cp ) {
+                // space
+                case 0x0020: result.append(genFullHalfMatchRegex((char)cp, (char)0x3000)); break;
```

Comment:
I disagree that a table lookup will be "simpler" in the end. Creating the table will take a similar number of lines, and then you will be wasting a lot of memory with persistent objects instead of inline `int`s. I agree that a long `switch` statement is ugly, but I think it's the most appropriate choice here.